### ISSUE TITLE
Trouble installing and running `predict_bulk_TF.py`

### ISSUE THREAD

**Issue by tacazares** [sperfu/scFAN#4]
Hello, 
I am excited to test out scFAN for bulk prediction, but I have had a few issues with install. I tried using your method for installing scFAN for python 3 with your conda approach described in the Intro Readme.md. I ran into a conflict between packages that did not let me install from your `scFAN.yml` file. 

```bash
(base) bash-4.2$ conda env create -f scFAN.yml
Collecting package metadata (repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.14.0
  latest version: 22.9.0

Please update conda by running

    $ conda update -n base -c conda-forge conda


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Installing pip dependencies: - Ran pip subprocess with arguments:
['/users/caz3so/.conda/envs/mamba/envs/scFAN2/bin/python', '-m', 'pip', 'install', '-U', '-r', '/data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt']
Pip subprocess output:
Collecting absl-py==0.15.0
  Using cached absl_py-0.15.0-py3-none-any.whl (132 kB)
Collecting astunparse==1.6.3
  Using cached astunparse-1.6.3-py2.py3-none-any.whl (12 kB)
Collecting cached-property==1.5.2
  Using cached cached_property-1.5.2-py2.py3-none-any.whl (7.6 kB)
Collecting cachetools==4.2.4
  Using cached cachetools-4.2.4-py3-none-any.whl (10 kB)
Collecting charset-normalizer==2.0.9
  Using cached charset_normalizer-2.0.9-py3-none-any.whl (39 kB)
Collecting clang==5.0
  Using cached clang-5.0.tar.gz (30 kB)
Collecting dataclasses==0.8
  Using cached dataclasses-0.8-py3-none-any.whl (19 kB)
Collecting flatbuffers==1.12
  Using cached flatbuffers-1.12-py2.py3-none-any.whl (15 kB)
Collecting gast==0.4.0
  Using cached gast-0.4.0-py3-none-any.whl (9.8 kB)
Collecting google-auth==1.35.0
  Using cached google_auth-1.35.0-py2.py3-none-any.whl (152 kB)
Collecting google-auth-oauthlib==0.4.6
  Using cached google_auth_oauthlib-0.4.6-py2.py3-none-any.whl (18 kB)
Requirement already satisfied: google-pasta==0.2.0 in /users/caz3so/.local/lib/python3.6/site-packages (from -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 12)) (0.2.0)
Collecting grpcio==1.43.0
  Using cached grpcio-1.43.0-cp36-cp36m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.1 MB)
Requirement already satisfied: h5py==2.10.0 in /users/caz3so/.local/lib/python3.6/site-packages (from -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 14)) (2.10.0)
Collecting idna==3.3
  Using cached idna-3.3-py3-none-any.whl (61 kB)
Collecting importlib-metadata==4.8.3
  Using cached importlib_metadata-4.8.3-py3-none-any.whl (17 kB)
Collecting keras==2.0.6
  Using cached Keras-2.0.6-py3-none-any.whl
Requirement already satisfied: keras-applications==1.0.8 in /users/caz3so/.local/lib/python3.6/site-packages (from -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 18)) (1.0.8)
Requirement already satisfied: keras-preprocessing==1.1.2 in /users/caz3so/.local/lib/python3.6/site-packages (from -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 19)) (1.1.2)
Collecting markdown==3.3.6
  Using cached Markdown-3.3.6-py3-none-any.whl (97 kB)
Collecting oauthlib==3.1.1
  Using cached oauthlib-3.1.1-py2.py3-none-any.whl (146 kB)
Requirement already satisfied: opt-einsum==3.3.0 in /users/caz3so/.local/lib/python3.6/site-packages (from -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 22)) (3.3.0)
Collecting parmap==1.5.3
  Using cached parmap-1.5.3-py2.py3-none-any.whl (12 kB)
Collecting protobuf==3.19.1
  Using cached protobuf-3.19.1-cp36-cp36m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.1 MB)
Collecting pyasn1==0.4.8
  Using cached pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)
Collecting pyasn1-modules==0.2.8
  Using cached pyasn1_modules-0.2.8-py2.py3-none-any.whl (155 kB)
Collecting pybedtools==0.7.10
  Using cached pybedtools-0.7.10.tar.gz (12.6 MB)
Requirement already satisfied: pybigwig==0.3.18 in /users/caz3so/.local/lib/python3.6/site-packages (from -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 28)) (0.3.18)
Collecting pyfasta==0.5.2
  Using cached pyfasta-0.5.2.tar.gz (19 kB)
Collecting pysam==0.18.0
  Using cached pysam-0.18.0-cp36-cp36m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.0 MB)
Collecting pyyaml==6.0
  Using cached PyYAML-6.0-cp36-cp36m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (603 kB)
Collecting requests==2.26.0
  Using cached requests-2.26.0-py2.py3-none-any.whl (62 kB)
Collecting requests-oauthlib==1.3.0
  Using cached requests_oauthlib-1.3.0-py2.py3-none-any.whl (23 kB)
Collecting rsa==4.8
  Using cached rsa-4.8-py3-none-any.whl (39 kB)
Collecting scipy==1.5.4
  Using cached scipy-1.5.4-cp36-cp36m-manylinux1_x86_64.whl (25.9 MB)
Collecting tensorboard==2.6.0
  Using cached tensorboard-2.6.0-py3-none-any.whl (5.6 MB)
Collecting tensorboard-data-server==0.6.1
  Using cached tensorboard_data_server-0.6.1-py3-none-manylinux2010_x86_64.whl (4.9 MB)
Collecting tensorboard-plugin-wit==1.8.0
  Using cached tensorboard_plugin_wit-1.8.0-py3-none-any.whl (781 kB)
Collecting tensorflow==2.6.2
  Using cached tensorflow-2.6.2-cp36-cp36m-manylinux2010_x86_64.whl (458.3 MB)
Collecting tensorflow-estimator==2.6.0
  Using cached tensorflow_estimator-2.6.0-py2.py3-none-any.whl (462 kB)
Requirement already satisfied: termcolor==1.1.0 in /users/caz3so/.local/lib/python3.6/site-packages (from -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 41)) (1.1.0)
Collecting theano==1.0.5
  Using cached Theano-1.0.5-py3-none-any.whl
Requirement already satisfied: typing-extensions==3.7.4.3 in /users/caz3so/.local/lib/python3.6/site-packages (from -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 43)) (3.7.4.3)
Collecting urllib3==1.26.7
  Using cached urllib3-1.26.7-py2.py3-none-any.whl (138 kB)
Collecting werkzeug==2.0.2
  Using cached Werkzeug-2.0.2-py3-none-any.whl (288 kB)
Requirement already satisfied: wrapt==1.12.1 in /users/caz3so/.local/lib/python3.6/site-packages (from -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 46)) (1.12.1)
Collecting zipp==3.6.0
  Using cached zipp-3.6.0-py3-none-any.whl (5.3 kB)
Requirement already satisfied: six in /users/caz3so/.local/lib/python3.6/site-packages (from absl-py==0.15.0->-r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 1)) (1.15.0)
Requirement already satisfied: wheel<1.0,>=0.23.0 in /users/caz3so/.local/lib/python3.6/site-packages (from astunparse==1.6.3->-r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 2)) (0.36.2)
Requirement already satisfied: setuptools>=40.3.0 in /users/caz3so/.local/lib/python3.6/site-packages (from google-auth==1.35.0->-r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 10)) (56.0.0)
Requirement already satisfied: numpy>=1.7 in /users/caz3so/.local/lib/python3.6/site-packages (from h5py==2.10.0->-r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 14)) (1.19.5)
Requirement already satisfied: certifi>=2017.4.17 in /users/caz3so/.conda/envs/mamba/envs/scFAN2/lib/python3.6/site-packages (from requests==2.26.0->-r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 32)) (2021.5.30)
INFO: pip is looking at multiple versions of tensorboard-plugin-wit to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of tensorboard-data-server to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of tensorboard to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of scipy to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of rsa to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of requests-oauthlib to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of requests to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyyaml to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pysam to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyfasta to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pybigwig to determine which version is compatible with other requirements. This could take a while.
Collecting pybigwig==0.3.18
  Using cached pyBigWig-0.3.18-cp36-cp36m-linux_x86_64.whl
INFO: pip is looking at multiple versions of pybedtools to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyasn1-modules to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyasn1 to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of protobuf to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of parmap to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of opt-einsum to determine which version is compatible with other requirements. This could take a while.
Collecting opt-einsum==3.3.0
  Using cached opt_einsum-3.3.0-py3-none-any.whl (65 kB)
INFO: pip is looking at multiple versions of oauthlib to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of markdown to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of keras-preprocessing to determine which version is compatible with other requirements. This could take a while.
Collecting keras-preprocessing==1.1.2
  Using cached Keras_Preprocessing-1.1.2-py2.py3-none-any.whl (42 kB)
INFO: pip is looking at multiple versions of tensorboard-plugin-wit to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of tensorboard-data-server to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of tensorboard to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of scipy to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of rsa to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of requests-oauthlib to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of requests to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyyaml to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pysam to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyfasta to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pybigwig to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of keras-applications to determine which version is compatible with other requirements. This could take a while.
Collecting keras-applications==1.0.8
  Using cached Keras_Applications-1.0.8-py3-none-any.whl (50 kB)
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: pip is looking at multiple versions of pybedtools to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyasn1-modules to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyasn1 to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of protobuf to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of parmap to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of opt-einsum to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of keras to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of importlib-metadata to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of idna to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of h5py to determine which version is compatible with other requirements. This could take a while.
Collecting h5py==2.10.0
  Using cached h5py-2.10.0-cp36-cp36m-manylinux1_x86_64.whl (2.9 MB)
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: pip is looking at multiple versions of oauthlib to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of markdown to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of keras-preprocessing to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of grpcio to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of google-pasta to determine which version is compatible with other requirements. This could take a while.
Collecting google-pasta==0.2.0
  Using cached google_pasta-0.2.0-py3-none-any.whl (57 kB)
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: pip is looking at multiple versions of keras-applications to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of google-auth-oauthlib to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of google-auth to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of gast to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flatbuffers to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of dataclasses to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of clang to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of charset-normalizer to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of <Python from Requires-Python> to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of cachetools to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of cached-property to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of astunparse to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of absl-py to determine which version is compatible with other requirements. This could take a while.

The conflict is caused by:
    The user requested h5py==2.10.0
    keras-applications 1.0.8 depends on h5py
    tensorflow 2.6.2 depends on h5py~=3.1.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict


Pip subprocess error:
ERROR: Cannot install -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 18), -r /data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/condaenv.g8qwjiqb.requirements.txt (line 39) and h5py==2.10.0 because these package versions have conflicting dependencies.
ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies

failed

CondaEnvException: Pip failed
```

I tried a different approach and installed the listed packages from the main page. I started with python3 and had some success. Unfortunately I ran into issues with the python3 scripts. They do not have the correct syntax or functions for python3. The functions `xrange` and `izip` are not in python3 and should be replaces with `range` and `zip`.

```bash
(scfan_py36) bash-4.2$ python predict_bulk_TF.py -i ../Data/GM12878/ -oc ../model/GM12878/
2022-10-12 16:35:12.732283: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/bzip2/1.0.6/lib:/usr/local/binutils/2.30.0/lib:/usr/local/gcc/9.3.0/lib:/usr/local/gcc/9.3.0/lib64:/usr/local/mpfr/4.0.0/lib:/usr/local/mpc/1.0.2/lib:/usr/local/gmp/6.0.0/lib:/data/weirauchlab/opt/lib:/data/weirauchlab/opt/lib64:/cm/shared/apps/lsf10/10.1/linux3.10-glibc2.17-x86_64/lib:/data/weirauchlab/opt/lib:/data/weirauchlab/opt/lib64:/data/weirauchlab/local/lib
2022-10-12 16:35:12.732369: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
Multi-task training
<_io.TextIOWrapper name='<stderr>' mode='w' encoding='utf-8'> output directory (../model/GM12878/) already exists so it will be clobbered
Loading genome
Loading ChIP labels
Sorting BED files
Removing peaks outside of X chromosome and autosomes
Windowing genome
Extracting windows that overlap at least one ChIP interval
Removing windows that overlap a blacklisted region
Number of positive windows: 843523
Number of targets: 15
Generating target matrix of all positive windows
Positive matrix sparsity 0.8698910798322432
Loading bigWig data
Making features
Splitting positive windows into training, validation, and testing sets
Traceback (most recent call last):
  File "predict_bulk_TF.py", line 276, in <module>
    main()
  File "predict_bulk_TF.py", line 227, in main
    utils.make_features_multiTask(positive_windows, y_positive,
  File "/data/miraldiNB/Tareian/scratch/20221010_scfan_gm12878/bin/scFAN/python3_codes/utils.py", line 85, in make_features_multiTask
    for positive_window, target_array in itertools.izip(positive_windows, y_positive):
AttributeError: module 'itertools' has no attribute 'izip'
```

I was able to make the modifications and run the script, but it seems that you must create all the training, validation, and testing data generators. Is there a simplified script that can perform only prediction? Additionally, I was getting an error about predict_generator() getting incorrect keyword arguments:

```bash
(scfan_py36) bash-4.2$ python predict_bulk_TF.py -i ../Data/GM12878 -oc ../model/GM12878
2022-10-12 18:01:56.807218: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/bzip2/1.0.6/lib:/usr/local/binutils/2.30.0/lib:/usr/local/gcc/9.3.0/lib:/usr/local/gcc/9.3.0/lib64:/usr/local/mpfr/4.0.0/lib:/usr/local/mpc/1.0.2/lib:/usr/local/gmp/6.0.0/lib:/data/weirauchlab/opt/lib:/data/weirauchlab/opt/lib64:/cm/shared/apps/lsf10/10.1/linux3.10-glibc2.17-x86_64/lib:/data/weirauchlab/opt/lib:/data/weirauchlab/opt/lib64:/data/weirauchlab/local/lib
2022-10-12 18:01:56.807267: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
Multi-task training
<_io.TextIOWrapper name='<stderr>' mode='w' encoding='utf-8'> output directory (../model/GM12878) already exists so it will be clobbered
Loading genome
Loading ChIP labels
Sorting BED files
Removing peaks outside of X chromosome and autosomes
Windowing genome
Extracting windows that overlap at least one ChIP interval
Removing windows that overlap a blacklisted region
Number of positive windows: 843523
Number of targets: 15
Generating target matrix of all positive windows
Positive matrix sparsity 0.8698910798322432
Loading bigWig data
Making features
Splitting positive windows into training, validation, and testing sets
Getting negative training examples
Getting negative validation examples
Getting negative testing examples
Extracting data from negative training BEDs
Extracting data from negative validation BEDs
Extracting data from negative testing BEDs
Shuffling training data
Generating data iterators
158899800 training samples
98048 validation samples
0 test samples
Building model
2022-10-12 18:25:37.229360: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/bzip2/1.0.6/lib:/usr/local/binutils/2.30.0/lib:/usr/local/gcc/9.3.0/lib:/usr/local/gcc/9.3.0/lib64:/usr/local/mpfr/4.0.0/lib:/usr/local/mpc/1.0.2/lib:/usr/local/gmp/6.0.0/lib:/data/weirauchlab/opt/lib:/data/weirauchlab/opt/lib64:/cm/shared/apps/lsf10/10.1/linux3.10-glibc2.17-x86_64/lib:/data/weirauchlab/opt/lib:/data/weirauchlab/opt/lib64:/data/weirauchlab/local/lib
2022-10-12 18:25:37.229441: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)
2022-10-12 18:25:37.229506: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (bmi-460g9-08.chmcres.cchmc.org): /proc/driver/nvidia/version does not exist
2022-10-12 18:25:37.230019: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Traceback (most recent call last):
  File "predict_bulk_TF.py", line 276, in <module>
    main()
  File "predict_bulk_TF.py", line 233, in main
    model_predicts,model_tfs = test(output_dir,datagen_valid)
  File "predict_bulk_TF.py", line 23, in test
    model_predicts = model.predict_generator(datagen_test, val_samples=1+len(datagen_test)/100, pickle_safe=True,verbose=1)
TypeError: predict_generator() got an unexpected keyword argument 'val_samples'
```

Have you ever had this issue? Thanks for your help!

**Comment by tacazares** [comment]
I was able to test installation on a different machine. I found that installing with conda seemed to cause a lot of conflicts. I was able to install the needed packages with pip. I had another issue though that caused the generator to fail. 

```bash
(scfan) bash-4.2$ python predict_bulk_TF.py  --inputdirs /fs/ess/PES0738/tareian/scfan/scFAN/Data/GM12878 -oc /fs/ess/PES0738/tareian/scfan/scFAN/model/GM12878
Using Theano backend.
Multi-task training
<_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'> output directory (/fs/ess/PES0738/tareian/scfan/scFAN/model/GM12878) already exists so it will be clobbered
Loading genome
Loading ChIP labels
Sorting BED files
Removing peaks outside of X chromosome and autosomes
Windowing genome
Extracting windows that overlap at least one ChIP interval
Removing windows that overlap a blacklisted region
Number of positive windows: 843523
Number of targets: 15
Generating target matrix of all positive windows
Positive matrix sparsity 0.8698910798322432
Loading bigWig data
Making features
Splitting positive windows into training, validation, and testing sets
Getting negative training examples
Getting negative validation examples
Getting negative testing examples
Extracting data from negative training BEDs
Extracting data from negative validation BEDs
Extracting data from negative testing BEDs
Shuffling training data
Generating data iterators
158899800 training samples
98048 validation samples
0 test samples
Building model
predict_bulk_TF.py:23: UserWarning: The semantics of the Keras 2 argument `steps_per_epoch` is not the same as the Keras 1 argument `samples_per_epoch`. `steps_per_epoch` is the number of batches to draw from the generator at each epoch. Basically steps_per_epoch = samples_per_epoch/batch_size. Similarly `nb_val_samples`->`validation_steps` and `val_samples`->`steps` arguments have changed. Update your method calls accordingly.
  model_predicts = model.predict_generator(datagen_test, val_samples=1+len(datagen_test)/100, pickle_safe=True,verbose=1)
predict_bulk_TF.py:23: UserWarning: Update your `predict_generator` call to the Keras 2 API: `predict_generator(<data_iter..., verbose=1, steps=981.48, use_multiprocessing=True)`
  model_predicts = model.predict_generator(datagen_test, val_samples=1+len(datagen_test)/100, pickle_safe=True,verbose=1)
Process Process-57:
Traceback (most recent call last):
  File "predict_bulk_TF.py", line 276, in <module>
    main()
  File "predict_bulk_TF.py", line 233, in main
    model_predicts,model_tfs = test(output_dir,datagen_valid)
  File "predict_bulk_TF.py", line 23, in test
    model_predicts = model.predict_generator(datagen_test, val_samples=1+len(datagen_test)/100, pickle_safe=True,verbose=1)
  File "/users/PES0738/ucn2855/.conda/envs/scfan/lib/python3.6/site-packages/keras/legacy/interfaces.py", line 87, in wrapper
    return func(*args, **kwargs)
  File "/users/PES0738/ucn2855/.conda/envs/scfan/lib/python3.6/site-packages/keras/engine/training.py", line 2067, in predict_generator
    generator_output = next(output_generator)
StopIteration
```

**Comment by sperfu** [comment]
Hi there, 

Thanks for your interest in our work. Regarding to the  syntax or functions errors, we have correct those errors in the codes. As for the predict_generator issue, I assume that it might either due to the settings on  the number of val_samples or the package version issue(either Keras or its backend Theano). You may first try use different version of backend framework(Theano or Tensorflow?), then Keras with different version. Then if it is not working, you may try to alter the setttings on the val_sample parameters.

Thanks.

**Comment by tacazares** [comment]
Hey, 
Thanks a lot for your help! I ended up using the python2 version with a few modifications. I will post some of the troubleshooting that I ended up doing below. I was able to change the backend used by Keras to Theano as suggested to fix my last problem. I used the information from [this post](https://data-flair.training/blogs/keras-backend/ ) to learn how to adjust my `keras.json` file. 

I also kept getting the error:

```bash
Building model
Traceback (most recent call last):
  File "predict_bulk_TF.py", line 274, in <module>
    main()
  File "predict_bulk_TF.py", line 231, in main
    model_predicts,model_tfs = test(output_dir,datagen_valid)
  File "predict_bulk_TF.py", line 20, in test
    model_tfs, model_bigwig_names, features, model = utils.load_model(model_dir)
  File "/fs/ess/PES0738/tareian/scfan/scFAN/utils.py", line 766, in load_model
    model = model_from_json(model_json)
  File "/users/PES0738/ucn2855/.conda/envs/scfan2/lib/python2.7/site-packages/keras/models.py", line 325, in model_from_json
    return layer_module.deserialize(config, custom_objects=custom_obje
cts)
  File "/users/PES0738/ucn2855/.conda/envs/scfan2/lib/python2.7/site-packages/keras/layers/__init__.py", line 46, in deserialize
    printable_module_name='layer')
  File "/users/PES0738/ucn2855/.conda/envs/scfan2/lib/python2.7/site-packages/keras/utils/generic_utils.py", line 140, in deserialize_keras_object
    list(custom_objects.items())))
  File "/users/PES0738/ucn2855/.conda/envs/scfan2/lib/python2.7/site-packages/keras/engine/topology.py", line 2374, in from_config
    process_layer(layer_data)
  File "/users/PES0738/ucn2855/.conda/envs/scfan2/lib/python2.7/site-packages/keras/engine/topology.py", line 2369, in process_layer
    layer(input_tensors[0], **kwargs)
  File "/users/PES0738/ucn2855/.conda/envs/scfan2/lib/python2.7/site-packages/keras/engine/topology.py", line 554, in __call__
    output = self.call(inputs, **kwargs)
  File "/users/PES0738/ucn2855/.conda/envs/scfan2/lib/python2.7/site-packages/keras/layers/normalization.py", line 188, in call
    training=training)
  File "/users/PES0738/ucn2855/.conda/envs/scfan2/lib/python2.7/site-packages/keras/backend/theano_backend.py", line 1333, in in_train_phase
    x = theano.ifelse.ifelse(training, x, alt)
AttributeError: 'module' object has no attribute 'ifelse'
```

 I just had to add the line `import theano.ifelse` to the file `.conda/envs/scfan2/lib/python2.7/site-packages/keras/backend/theano_backend.py`. Hope this helps others who have trouble with conda. 
Tareian

**Comment by gurudatta-patil** [comment]
```Multi-task training
output directory (mul) already exists so it will be clobbered
Loading genome
Loading ChIP labels
Sorting BED files
Removing peaks outside of X chromosome and autosomes
Windowing genome
Extracting windows that overlap at least one ChIP interval
Removing windows that overlap a blacklisted region
Number of positive windows: 859619
Number of targets: 15
Generating target matrix of all positive windows
Positive matrix sparsity 0.8697287209023222
Loading bigWig data
Making features
Splitting positive windows into training, validation, and testing sets
Getting negative training examples
Getting negative validation examples
Getting negative testing examples
Extracting data from negative training BEDs
Extracting data from negative validation BEDs
Extracting data from negative testing BEDs
Shuffling training data
Generating data iterators
162086400 training samples
98374 validation samples
0 test samples
Building model
predict_bulk_TF.py:27: UserWarning: Update your `predict_generator` call to the Keras 2 API: `predict_generator(<data_iter..., pickle_safe=True, steps=984, verbose=1)`
  model_predicts = model.predict_generator(datagen_test, val_samples=1+len(datagen_test)/100, pickle_safe=True,verbose=1)
[urlOpen] Couldn't open data/GM12878/wgEncodeDukeMapabilityUniqueness35bp.bigWig for reading
[urlOpen] Couldn't open data/GM12878/wgEncodeDukeMapabilityUniqueness35bp.bigWig for reading
[pyBwOpen] bw is NULL!
Process Process-49:
Traceback (most recent call last):
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
Traceback (most recent call last):
  File "predict_bulk_TF.py", line 279, in <module>
    main()
  File "predict_bulk_TF.py", line 236, in main
    model_predicts,model_tfs = test(output_dir,datagen_valid)
  File "predict_bulk_TF.py", line 27, in test
    model_predicts = model.predict_generator(datagen_test, val_samples=1+len(datagen_test)/100, pickle_safe=True,verbose=1)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/legacy/interfaces.py", line 88, in wrapper
    return func(*args, **kwargs)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/engine/training.py", line 2094, in predict_generator
    outs = self.predict_on_batch(x)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/engine/training.py", line 1683, in predict_on_batch
    outputs = self.predict_function(ins)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/backend/theano_backend.py", line 1095, in __call__
    return self.function(*inputs)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/theano/compile/function_module.py", line 917, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/theano/gof/link.py", line 325, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/theano/compile/function_module.py", line 903, in __call__
    self.fn() if output_subset is None else\
NotImplementedError: input nd
Apply node that caused the error: InplaceDimShuffle{0,3,1,2}(/input_1)
Toposort index: 2
Inputs types: [TensorType(float32, 4D)]
Inputs shapes: ['No shapes']
Inputs strides: ['No strides']
Inputs values: [None]
Outputs clients: [[CorrMM{half, (1, 1), (1, 1), 1 False}(InplaceDimShuffle{0,3,1,2}.0, Subtensor{::, ::, ::int64, ::int64}.0)]]

Backtrace when the node is created(use Theano flag traceback.limit=N to make it longer):
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/layers/__init__.py", line 46, in deserialize
    printable_module_name='layer')
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/utils/generic_utils.py", line 140, in deserialize_keras_object
    list(custom_objects.items())))
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/engine/topology.py", line 2374, in from_config
    process_layer(layer_data)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/engine/topology.py", line 2369, in process_layer
    layer(input_tensors[0], **kwargs)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/engine/topology.py", line 554, in __call__
    output = self.call(inputs, **kwargs)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/layers/convolutional.py", line 164, in call
    dilation_rate=self.dilation_rate)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/backend/theano_backend.py", line 1718, in conv2d
    x = _preprocess_conv2d_input(x, data_format)
  File "/home/gvpatil/miniconda3/envs/scfan2/lib/python2.7/site-packages/keras/backend/theano_backend.py", line 1494, in _preprocess_conv2d_input
    x = x.dimshuffle((0, 3, 1, 2))

HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.
```
Any idea why this error occurs while running the python 2 version of the code?
